diff --git a/subprojects/gst-plugins-good/sys/v4l2/gstv4l2.c b/subprojects/gst-plugins-good/sys/v4l2/gstv4l2.c
index ec53bc9e83..25cbeace71 100644
--- a/subprojects/gst-plugins-good/sys/v4l2/gstv4l2.c
+++ b/subprojects/gst-plugins-good/sys/v4l2/gstv4l2.c
@@ -57,6 +57,7 @@
 #include "gstv4l2vp8enc.h"
 #include "gstv4l2vp9enc.h"
 #include "gstv4l2transform.h"
+#include "hailo_vsm/hailo_vsm_meta.h"
 
 GST_DEBUG_CATEGORY_EXTERN (v4l2_debug);
 #define GST_CAT_DEFAULT v4l2_debug
@@ -267,6 +268,9 @@ plugin_init (GstPlugin * plugin)
   ret |= GST_ELEMENT_REGISTER (v4l2radio, plugin);
   ret |= GST_DEVICE_PROVIDER_REGISTER (v4l2deviceprovider, plugin);
 
+  gst_hailo_vsm_meta_get_info();
+  gst_hailo_vsm_meta_get_type();
+
   return ret;
 }
 
diff --git a/subprojects/gst-plugins-good/sys/v4l2/gstv4l2src.c b/subprojects/gst-plugins-good/sys/v4l2/gstv4l2src.c
index d00f7a80dc..e8ec2ce1c6 100644
--- a/subprojects/gst-plugins-good/sys/v4l2/gstv4l2src.c
+++ b/subprojects/gst-plugins-good/sys/v4l2/gstv4l2src.c
@@ -57,6 +57,10 @@
 #include "gstv4l2elements.h"
 #include "gstv4l2src.h"
 
+#include "hailo_vsm/hailo_vsm_meta.h"
+#include "hailo_vsm/hailo_vsm.h"
+#include "gst/allocators/gstdmabuf.h"
+
 #include "gstv4l2colorbalance.h"
 #include "gstv4l2tuner.h"
 #include "gstv4l2vidorient.h"
@@ -67,6 +71,7 @@ GST_DEBUG_CATEGORY (v4l2src_debug);
 #define GST_CAT_DEFAULT v4l2src_debug
 
 #define DEFAULT_PROP_DEVICE   "/dev/video0"
+#define VIDEO_GET_VSM_IOC 0x1337face
 
 enum
 {
@@ -924,6 +929,61 @@ gst_v4l2src_change_state (GstElement * element, GstStateChange transition)
   return ret;
 }
 
+static gboolean
+gst_v4l2_get_vsm_metadata(GstV4l2Src *v4l2src, GstV4l2Object *obj, GstBuffer *buf)
+{
+    struct hailo15_get_vsm_params vsm_params;
+    guint v4l2_index;
+
+    GstMemory *mem;
+    GstV4l2Memory *vmem;
+    GstV4l2MemoryGroup *group;      
+
+    // Get V4l2 memory group from buffer - taken from gstv4l2bufferpool.c (gst_v4l2_is_buffer_valid)
+    mem = gst_buffer_peek_memory(buf, 0);
+
+    if (GST_BUFFER_FLAG_IS_SET(buf, GST_BUFFER_FLAG_TAG_MEMORY))
+    {
+      GST_WARNING_OBJECT(v4l2src, "Buffer has tag memory. skip adding VSM metadata");
+      return TRUE;
+    }
+
+    if (gst_is_dmabuf_memory (mem))
+      mem = gst_mini_object_get_qdata (GST_MINI_OBJECT (mem),
+          GST_V4L2_MEMORY_QUARK);
+
+    if(!mem)
+    {
+      GST_WARNING_OBJECT(v4l2src, "Failed to get memory from buffer. skip adding VSM metadata");
+      return TRUE;
+    }
+
+    if (!gst_is_v4l2_memory(mem)){
+      GST_WARNING_OBJECT(v4l2src, "Failed to get v4l2memory from buffer (buffer may copied). skip adding VSM metadata");
+      return TRUE;
+    }
+
+    vmem = (GstV4l2Memory *) mem;
+    group = vmem->group;
+
+    // Get the v4l2 buffer index from the memory group
+    v4l2_index = group->buffer.index;
+    vsm_params.index = v4l2_index;
+
+    GST_DEBUG_OBJECT(v4l2src, "v4l2 index is %d, Perform IOCTL", v4l2_index);
+    // Get the VSM params (motion vector) from the video device
+    if(obj->ioctl(obj->video_fd, VIDEO_GET_VSM_IOC, &vsm_params)){
+      GST_WARNING_OBJECT(v4l2src, "Failed to get vsm params for index %d", v4l2_index);
+      return TRUE;
+    }
+
+    GST_DEBUG_OBJECT(v4l2src, "Got vsm params for index %d, x = %d , y=%d", v4l2_index, vsm_params.vsm.dx, vsm_params.vsm.dy);
+
+    // Add the VSM params to the buffer as metadata
+    gst_buffer_add_hailo_vsm_meta(buf, v4l2_index, vsm_params.vsm);
+    return TRUE;
+}
+
 static GstFlowReturn
 gst_v4l2src_create (GstPushSrc * src, GstBuffer ** buf)
 {
@@ -973,6 +1033,9 @@ gst_v4l2src_create (GstPushSrc * src, GstBuffer ** buf)
   if (G_UNLIKELY (ret != GST_FLOW_OK))
     goto error;
 
+  if (!gst_v4l2_get_vsm_metadata(v4l2src, obj, *buf))
+    goto error;
+
   timestamp = GST_BUFFER_TIMESTAMP (*buf);
   duration = obj->duration;
 
diff --git a/subprojects/gst-plugins-good/sys/v4l2/hailo_vsm/hailo_vsm.h b/subprojects/gst-plugins-good/sys/v4l2/hailo_vsm/hailo_vsm.h
new file mode 100644
index 0000000000..f93334497c
--- /dev/null
+++ b/subprojects/gst-plugins-good/sys/v4l2/hailo_vsm/hailo_vsm.h
@@ -0,0 +1,12 @@
+#pragma once
+// hailo_vsm name
+
+struct hailo15_vsm{
+  int dx;
+  int dy;
+};
+
+struct hailo15_get_vsm_params{
+  int index;
+  struct hailo15_vsm vsm;
+};
\ No newline at end of file
diff --git a/subprojects/gst-plugins-good/sys/v4l2/hailo_vsm/hailo_vsm_meta.c b/subprojects/gst-plugins-good/sys/v4l2/hailo_vsm/hailo_vsm_meta.c
new file mode 100644
index 0000000000..7002f5e502
--- /dev/null
+++ b/subprojects/gst-plugins-good/sys/v4l2/hailo_vsm/hailo_vsm_meta.c
@@ -0,0 +1,153 @@
+/**
+ * Copyright (c) 2021-2022 Hailo Technologies Ltd. All rights reserved.
+ * Distributed under the LGPL license (https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt)
+ **/
+#include "hailo_vsm_meta.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+
+static gboolean gst_hailo_vsm_meta_init(GstMeta *meta, gpointer params, GstBuffer *buffer);
+static void gst_hailo_vsm_meta_free(GstMeta *meta, GstBuffer *buffer);
+static gboolean gst_hailo_vsm_meta_transform(GstBuffer *transbuf, GstMeta *meta, GstBuffer *buffer,
+                                         GQuark type, gpointer data);
+
+// Register metadata type and returns Gtype
+// https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gstreamer/html/gstreamer-GstMeta.html#gst-meta-api-type-register
+GType gst_hailo_vsm_meta_get_type(void)
+{
+    static const gchar *tags[] = {NULL};
+    static GType type;
+    if (g_once_init_enter(&type))
+    {
+        GType _type = gst_meta_api_type_register("GstHailoVsmMetaAPI", tags);
+        g_once_init_leave(&type, _type);
+    }
+    return type;
+}
+
+// GstMetaInfo provides info for specific metadata implementation
+// https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gstreamer/html/gstreamer-GstMeta.html#GstMetaInfo
+const GstMetaInfo *gst_hailo_vsm_meta_get_info(void)
+{
+    static const GstMetaInfo *gst_hailo_vsm_meta_info = NULL;
+
+    if (g_once_init_enter(&gst_hailo_vsm_meta_info))
+    {
+        // Explanation of fields
+        // https://gstreamer.freedesktop.org/documentation/design/meta.html#gstmeta1
+        const GstMetaInfo *meta = gst_meta_register(GST_HAILO_VSM_META_API_TYPE, /* api type */
+                                                    "GstHailoVsmMeta",          /* implementation type */
+                                                    sizeof(GstHailoVsmMeta),    /* size of the structure */
+                                                    gst_hailo_vsm_meta_init,
+                                                    (GstMetaFreeFunction)gst_hailo_vsm_meta_free,
+                                                    gst_hailo_vsm_meta_transform);
+        g_once_init_leave(&gst_hailo_vsm_meta_info, meta);
+    }
+    return gst_hailo_vsm_meta_info;
+}
+
+// Meta init function
+// Fourth field in GstMetaInfo
+// https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gstreamer/html/gstreamer-GstMeta.html#GstMetaInitFunction
+static gboolean gst_hailo_vsm_meta_init(GstMeta *meta, gpointer params, GstBuffer *buffer)
+{
+    GstHailoVsmMeta *gst_hailo_vsm_meta = (GstHailoVsmMeta *)meta;
+    // GStreamer is allocating the GstHailoVsmMeta struct with POD allocation (like malloc) when
+    // it holds non POD type (shared_ptr). The memset assures there is no garbage data in this address.
+    // This is a temporary solution because memset to non POD type is undefined behavior.
+    // https://stackoverflow.com/questions/59747240/is-it-okay-to-memset-a-struct-which-has-an-another-struct-with-smart-pointer-mem?rq=1
+    // Opened an issue to replace this line with right initialization - MAD-1158.
+    gst_hailo_vsm_meta->v4l2_index = 0;
+    gst_hailo_vsm_meta->vsm.dx = 0;
+    gst_hailo_vsm_meta->vsm.dy = 0;
+    return TRUE;
+}
+
+// Meta free function
+// Fifth field in GstMetaInfo
+// https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gstreamer/html/gstreamer-GstMeta.html#GstMetaFreeFunction
+static void gst_hailo_vsm_meta_free(GstMeta *meta, GstBuffer *buffer)
+{
+    GstHailoVsmMeta *hailo_vsm_meta = (GstHailoVsmMeta *)meta;
+    hailo_vsm_meta->v4l2_index = 0;
+    hailo_vsm_meta->vsm.dx = 0;
+    hailo_vsm_meta->vsm.dy = 0;
+}
+
+// Meta transform function
+// Sixth field in GstMetaInfo
+// https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gstreamer/html/gstreamer-GstMeta.html#GstMetaTransformFunction
+static gboolean gst_hailo_vsm_meta_transform(GstBuffer *transbuf, GstMeta *meta, GstBuffer *buffer,
+                                             GQuark type, gpointer data)
+{
+    GstHailoVsmMeta *gst_hailo_vsm_meta = (GstHailoVsmMeta *)meta;
+
+    GstHailoVsmMeta *new_hailo_vsm_meta = gst_buffer_add_hailo_vsm_meta(transbuf, gst_hailo_vsm_meta->v4l2_index, gst_hailo_vsm_meta->vsm);
+    if(!new_hailo_vsm_meta)
+    {
+        GST_ERROR("gst_hailo_vsm_meta_transform: failed to transform hailo_vsm_meta");
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+GstHailoVsmMeta *gst_buffer_get_hailo_vsm_meta(GstBuffer *buffer)
+{
+    // https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gstreamer/html/GstBuffer.html#gst-buffer-get-meta
+    GstHailoVsmMeta *meta = (GstHailoVsmMeta *)gst_buffer_get_meta((buffer), GST_HAILO_VSM_META_API_TYPE);
+    return meta;
+}
+/**
+ * @brief Addes a new GstHailoVsmMeta to a given buffer, this meta is initialized with a given index.
+ *
+ * @param buffer Buffer to add the metadata on.
+ * @param guint v4l2_index
+ * @return GstHailoVsmMeta* The meta structure that was added to the buffer.
+ */
+GstHailoVsmMeta *gst_buffer_add_hailo_vsm_meta(GstBuffer *buffer, guint v4l2_index, struct hailo15_vsm vsm)
+{
+    GstHailoVsmMeta *gst_hailo_vsm_meta = NULL;
+
+    // check that gst_buffer valid
+    g_return_val_if_fail((int)GST_IS_BUFFER(buffer), NULL);
+
+    // check that gst_buffer writable
+    if (!gst_buffer_is_writable(buffer))
+        return gst_hailo_vsm_meta;
+
+    // https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gstreamer/html/GstBuffer.html#gst-buffer-add-meta
+    gst_hailo_vsm_meta = (GstHailoVsmMeta *)gst_buffer_add_meta(buffer, GST_HAILO_VSM_META_INFO, NULL);
+
+    gst_hailo_vsm_meta->v4l2_index = v4l2_index;
+    gst_hailo_vsm_meta->vsm = vsm;
+
+    return gst_hailo_vsm_meta;
+}
+
+/**
+ * @brief  Removes GstHailoVsmMeta from a given buffer.
+ *
+ * @param buffer A buffer to remove meta from.
+ * @return gboolean whether removal was successfull (TRUE if there isn't GstHailoVsmMeta).
+ * @note Removes only the first GstHailoVsmMeta in this buffer.
+ */
+gboolean gst_buffer_remove_hailo_vsm_meta(GstBuffer *buffer)
+{
+    GstHailoVsmMeta *meta;
+    g_return_val_if_fail((int)GST_IS_BUFFER(buffer), false);
+
+    meta = (GstHailoVsmMeta *)gst_buffer_get_meta((buffer), GST_HAILO_VSM_META_API_TYPE);
+
+    if (meta == NULL)
+        return TRUE;
+
+    if (!gst_buffer_is_writable(buffer))
+        return FALSE;
+
+    // https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gstreamer/html/GstBuffer.html#gst-buffer-remove-meta
+    return gst_buffer_remove_meta(buffer, &meta->meta);
+}
\ No newline at end of file
diff --git a/subprojects/gst-plugins-good/sys/v4l2/hailo_vsm/hailo_vsm_meta.h b/subprojects/gst-plugins-good/sys/v4l2/hailo_vsm/hailo_vsm_meta.h
new file mode 100644
index 0000000000..8661fcc1ad
--- /dev/null
+++ b/subprojects/gst-plugins-good/sys/v4l2/hailo_vsm/hailo_vsm_meta.h
@@ -0,0 +1,41 @@
+#include <gst/gst.h>
+#include "hailo_vsm.h"
+
+G_BEGIN_DECLS
+
+// Api Type
+// First field of gst_meta_register (which returns GstMetaInfo)
+// https://gstreamer.freedesktop.org/documentation/gstreamer/gstmeta.html?gi-language=c#gst_meta_register
+#define GST_HAILO_VSM_META_API_TYPE (gst_hailo_vsm_meta_get_type())
+#define GST_HAILO_VSM_META_INFO (gst_hailo_vsm_meta_get_info())
+
+#define HAILO_VSM_META_API_NAME "GstHailoVsmMetaAPI"
+
+typedef struct _GstHailoVsmMeta GstHailoVsmMeta;
+
+struct _GstHailoVsmMeta
+{
+    // Gstreamer Metadata for passing the VSM (motion vector) through hailo15 pipeline
+
+    GstMeta meta;
+    // V4L2 buffer index
+    guint v4l2_index;
+    // VSM
+    struct hailo15_vsm vsm;
+};
+
+GType gst_hailo_vsm_meta_get_type(void);
+
+GST_EXPORT
+const GstMetaInfo *gst_hailo_vsm_meta_get_info(void);
+
+GST_EXPORT
+GstHailoVsmMeta *gst_buffer_add_hailo_vsm_meta(GstBuffer *buffer, guint v4l2_index, struct hailo15_vsm vsm);
+
+GST_EXPORT
+gboolean gst_buffer_remove_hailo_vsm_meta(GstBuffer *buffer);
+
+GST_EXPORT
+GstHailoVsmMeta *gst_buffer_get_hailo_vsm_meta(GstBuffer *b);
+
+G_END_DECLS
\ No newline at end of file
diff --git a/subprojects/gst-plugins-good/sys/v4l2/meson.build b/subprojects/gst-plugins-good/sys/v4l2/meson.build
index 65f551fb64..bfde728cad 100644
--- a/subprojects/gst-plugins-good/sys/v4l2/meson.build
+++ b/subprojects/gst-plugins-good/sys/v4l2/meson.build
@@ -33,7 +33,8 @@ v4l2_sources = [
   'v4l2-utils.c',
   'tuner.c',
   'tunerchannel.c',
-  'tunernorm.c'
+  'tunernorm.c',
+  'hailo_vsm/hailo_vsm_meta.c'
 ]
 
 v4l2 = get_option('v4l2')
